File: demo_files/vulnerable_example.js
/**
 * Vulnerable JavaScript/Node.js code - for demo purposes only
 * This file intentionally contains security vulnerabilities
 * to demonstrate VulAI's detection and refactoring capabilities.
 */

const express = require('express');
const sqlite3 = require('sqlite3');
const { exec } = require('child_process');
const fs = require('fs');

const app = express();
app.use(express.json());

// Vulnerability 1: Hardcoded Credentials
const DB_USERNAME = 'admin';
const DB_PASSWORD = 'password123';
const SECRET_KEY = 'sk-test-1234567890abcdefghijklmno';

// Vulnerability 2: SQL Injection
app.get('/api/user/:id', (req, res) => {
  const userId = req.params.id;
  const db = new sqlite3.Database(':memory:');

  // VULNERABLE: Direct string concatenation in SQL
  const query = `SELECT * FROM users WHERE id = ${userId}`;

  db.all(query, (err, rows) => {
    if (err) {
      res.status(500).json({ error: err.message });
    } else {
      res.json(rows);
    }
  });
});

// Vulnerability 3: Command Injection
app.post('/api/process-file', (req, res) => {
  const filename = req.body.filename;

  // VULNERABLE: User input directly in exec() command
  exec(`file ${filename}`, (error, stdout, stderr) => {
    if (error) {
      res.status(500).json({ error: error.message });
    }
    res.json({ result: stdout });
  });
});

// Vulnerability 4: Path Traversal
app.get('/api/file/:path', (req, res) => {
  const filepath = req.params.path;

  // VULNERABLE: No validation allows ../../etc/passwd type attacks
  fs.readFile(`/data/${filepath}`, 'utf8', (err, data) => {
    if (err) {
      res.status(500).json({ error: err.message });
    }
    res.json({ content: data });
  });
});

// Vulnerability 5: Code Injection via eval
app.post('/api/calc', (req, res) => {
  const expression = req.body.expression;

  // VULNERABLE: eval() executes arbitrary code
  try {
    const result = eval(expression);
    res.json({ result });
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
});

// Vulnerability 6: Authentication Bypass
app.post('/api/login', (req, res) => {
  const username = req.body.username;
  const password = req.body.password;
  const db = new sqlite3.Database(':memory:');

  // VULNERABLE: SQL injection + weak authentication
  const query = `SELECT * FROM users WHERE username = '${username}' AND password = '${password}'`;

  db.get(query, (err, user) => {
    if (user) {
      // VULNERABLE: Storing password in session/response
      res.json({
        authenticated: true,
        user: user,
        password: password  // Never do this!
      });
    } else {
      res.json({ authenticated: false });
    }
  });
});

// Vulnerability 7: Insecure Cookie Usage
app.post('/api/session', (req, res) => {
  const userId = req.body.userId;

  // VULNERABLE: Insecure cookie - should have httpOnly, secure, sameSite flags
  res.cookie('sessionId', userId, {
    // Missing httpOnly, secure, sameSite flags
    expires: new Date(Date.now() + 3600000)
  });

  res.json({ sessionId: userId });
});

// Vulnerability 8: Prototype Pollution
app.post('/api/update-user', (req, res) => {
  const userData = req.body;
  const user = { name: '', email: '' };

  // VULNERABLE: Direct object assignment allows prototype pollution
  for (const key in userData) {
    user[key] = userData[key];  // __proto__, constructor allowed!
  }

  res.json({ user });
});

// Vulnerability 9: Logging Sensitive Data
app.use((req, res, next) => {
  // VULNERABLE: Logging entire request/response including credentials
  console.log('Request:', JSON.stringify(req.body));
  console.log('Headers:', req.headers);
  console.log('Authorization:', req.headers.authorization);
  next();
});

// Vulnerability 10: Unsafe Redirect
app.get('/api/redirect', (req, res) => {
  const url = req.query.url;

  // VULNERABLE: No validation of redirect URL - open redirect
  res.redirect(url);
});

// Vulnerability 11: Insecure Deserialization
app.post('/api/deserialize', (req, res) => {
  const serializedData = req.body.data;

  // VULNERABLE: eval() on user input (double vulnerability)
  const data = eval(`(${serializedData})`);

  res.json({ deserialized: data });
});

// Vulnerability 12: Missing Security Headers
app.get('/api/resource', (req, res) => {
  // VULNERABLE: No security headers set
  res.json({ data: 'sensitive' });
  // Missing: Content-Security-Policy, X-Frame-Options, X-Content-Type-Options, etc.
});

// Vulnerability 13: Race Condition in File Upload
const uploadedFiles = new Map();

app.post('/api/upload', (req, res) => {
  const filename = req.body.filename;
  const content = req.body.content;

  // VULNERABLE: No locking - multiple uploads can overwrite each other
  if (uploadedFiles.has(filename)) {
    return res.status(409).json({ error: 'File exists' });
  }

  // VULNERABLE: Time of check - time of use (TOCTOU) race condition
  uploadedFiles.set(filename, content);

  res.json({ uploaded: filename });
});

// Start server - VULNERABLE: No HTTPS, insecure configuration
app.listen(3000, '0.0.0.0', () => {
  console.log('Server listening on port 3000');
  console.log('Database password:', DB_PASSWORD);  // Logging secrets!
});

File: demo_files/vulnerable_example.py
"""
Vulnerable Python code - for demo purposes only
This script intentionally contains security vulnerabilities
to demonstrate VulAI's detection and refactoring capabilities.
"""

import os
import sqlite3
import subprocess
from flask import Flask, request

app = Flask(__name__)

# Vulnerability 1: Hardcoded Secret
API_KEY = "sk-proj-1234567890abcdefghijklmnop"
DATABASE_PASSWORD = "admin123"

# Vulnerability 2: Unsafe Database Query (SQL Injection)
def get_user_by_id(user_id):
    """Get user from database - vulnerable to SQL injection"""
    db = sqlite3.connect(':memory:')
    # VULNERABLE: Direct string concatenation in SQL query
    query = f"SELECT * FROM users WHERE id = {user_id}"
    results = db.execute(query).fetchall()
    return results


# Vulnerability 3: Command Injection
def process_file(filename):
    """Process a file - vulnerable to command injection"""
    # VULNERABLE: User input directly in os.system()
    command = f"file {filename}"
    os.system(command)

    # VULNERABLE: Subprocess with shell=True
    result = subprocess.run(f"ls -la {filename}", shell=True, capture_output=True)
    return result.stdout


# Vulnerability 4: Code Injection
@app.route('/eval', methods=['POST'])
def unsafe_eval():
    """Evaluate code from user input - EXTREMELY DANGEROUS"""
    user_code = request.json.get('code', '')

    # VULNERABLE: Direct eval of user input
    result = eval(user_code)
    return {'result': result}


# Vulnerability 5: Missing Input Validation
@app.route('/login', methods=['POST'])
def login():
    """Login endpoint without proper validation"""
    username = request.json.get('username')
    password = request.json.get('password')

    # No validation - directly used in query (SQL injection)
    db = sqlite3.connect(':memory:')
    query = f"SELECT * FROM users WHERE username = '{username}' AND password = '{password}'"
    user = db.execute(query).fetchone()

    if user:
        return {'authenticated': True, 'user': user}
    return {'authenticated': False}


# Vulnerability 6: Insecure Logging
@app.route('/process', methods=['POST'])
def process_request():
    """Process request - logs sensitive data"""
    api_key = request.headers.get('Authorization')
    user_data = request.json

    # VULNERABLE: Logging sensitive information
    print(f"Processing request from user: {user_data}")
    print(f"API Key: {api_key}")  # Never log API keys!

    return {'status': 'processed'}


# Vulnerability 7: Unsafe File Operations
def read_user_file(filename):
    """Read user-specified file - path traversal vulnerability"""
    # VULNERABLE: No validation of filename - allows path traversal
    with open(f'/data/{filename}', 'r') as f:
        return f.read()


# Vulnerability 8: Race Condition
def transfer_funds(from_account, to_account, amount):
    """Transfer funds between accounts - vulnerable to race condition"""
    db = sqlite3.connect(':memory:')

    # VULNERABLE: Check-then-act without locking
    balance = db.execute(f"SELECT balance FROM accounts WHERE id = {from_account}").fetchone()

    if balance[0] >= amount:
        # Multiple threads could execute between these lines
        db.execute(f"UPDATE accounts SET balance = balance - {amount} WHERE id = {from_account}")
        db.execute(f"UPDATE accounts SET balance = balance + {amount} WHERE id = {to_account}")
        db.commit()
        return True
    return False


if __name__ == '__main__':
    # VULNERABLE: Running in debug mode with host=0.0.0.0
    app.run(debug=True, host='0.0.0.0', port=5000)
